crates/shielder-scheduler-server/src/storage/dynamo_db.rs

```diff
@@ -0,0 +1,475 @@
+    }
+}
+
+impl StorageInterface for DynamoDb {
+    async fn insert_scheduled_request(
+        &self,
+        request: ScheduledRequest,
+    ) -> Result<(), StorageError> {
+        let payload = serde_json::to_string(&request)
+            .map_err(|e| StorageError::Internal(format!("Failed to serialize request: {e}")))?;
+
+        let mut builder = self
+            .client
+            .put_item()
+            .table_name(TABLE_NAME)
+            .condition_expression("attribute_not_exists(id)")
+            .item("id", AttributeValue::S(request.id.to_string()))
+            .item(
+                "status",
+                AttributeValue::S(status_to_str(&request.status).to_string()),
+            )
+            .item(
+                "relay_after",
+                AttributeValue::N(request.relay_after.timestamp().to_string()),
+            )
+            .item(
+                "created_at",
+                AttributeValue::N(request.created_at.timestamp().to_string()),
+            )
+            .item(
+                "last_note_index",
+                AttributeValue::S(request.last_note_index.to_string()),
+            )
+            .item("payload", AttributeValue::S(payload));
+
+        if let Some(err) = &request.error_message {
+            builder = builder.item("error_message", AttributeValue::S(err.clone()));
+        }
+        if let Some(processed_at) = request.processed_at {
+            builder = builder.item(
+                "processed_at",
+                AttributeValue::N(processed_at.timestamp().to_string()),
+            );
+        }
+
+        match builder.send().await {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                if is_conditional_check_failed(&e) {
+                    Err(StorageError::DuplicateEntry(request.id))
+                } else {
+                    Err(map_internal("put_item", e))
+                }
+            }
+        }
+    }
+
+    async fn get_pending_requests(
+        &self,
+        limit: usize,
+    ) -> Result<Vec<ScheduledRequest>, StorageError> {
+        let now = Utc::now().timestamp().to_string();
+        let mut all_requests = Vec::new();
+
+        // Query pending requests
+        let pending_requests = self
+            .query_requests_by_status_and_time(RequestStatus::Pending, &now, limit)
+            .await?;
+        all_requests.extend(pending_requests);
+
+        // Query processing requests if we need more
+        if all_requests.len() < limit {
+            let remaining_limit = limit - all_requests.len();
+            let processing_requests = self
+                .query_requests_by_status_and_time(RequestStatus::Processing, &now, remaining_limit)
+                .await?;
+            all_requests.extend(processing_requests);
+        }
+
+        // Sort by relay_after and take limit
+        all_requests.sort_by(|a, b| a.relay_after.cmp(&b.relay_after));
+        Ok(all_requests.into_iter().take(limit).collect())
```

Comment:
Ah you take limit here, so this is okay, but still I'm not sure second call to `query_requests_by_status_and_time` is correct.